### PR TITLE
KAFKA-4306: Shutdown distributed herder with a timeout.

### PR DESCRIPTION
**Pull request by kkonstantine**
Resolves

KAFKA-4306: Connect workers won't shut down if brokers are not available
KAFKA-4154: Kafka Connect fails to shutdown if it has not completed startup

### CODE REVIEW

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedHerder.java [hunk 1]
```diff
@@ -156,8 +159,16 @@ public DistributedHerder(DistributedConfig config,
...
         this.workerGroupId = config.getString(DistributedConfig.GROUP_ID_CONFIG);
         this.workerSyncTimeoutMs = config.getInt(DistributedConfig.WORKER_SYNC_TIMEOUT_MS_CONFIG);
+        this.workerTasksShutdownTimeoutMs = config.getLong(DistributedConfig.TASK_SHUTDOWN_GRACEFUL_TIMEOUT_MS_CONFIG);
         this.workerUnsyncBackoffMs = config.getInt(DistributedConfig.WORKER_UNSYNC_BACKOFF_MS_CONFIG);
         this.member = member != null ? member : new WorkerGroupMember(config, restUrl, this.configBackingStore, new RebalanceListener(), time);
+        this.herderExecutor = new ThreadPoolExecutor(1, 1, 0L, TimeUnit.MILLISECONDS, new LinkedBlockingDeque<Runnable>(1),
+                new ThreadFactory() {
+                    @Override
+                    public Thread newThread(Runnable herder) {
+                        return new Thread(herder, "DistributedHerder");
```
Comment (shikhar) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedHerder.java:169]: Should we also make them daemon threads?

**Comment by shikhar** [comment]
LGTM

**Comment by shikhar** [comment]
ping @ewencp

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedHerder.java (hunk 1 above)
Comment (kkonstantine) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedHerder.java:169]: I don't think this is necessary. Normal termination passes through `stop` method below, which first tries to shut them down gracefully, and after a timeout it shuts down these threads immediately. Abrupt terminations of the JVM will terminate those threads too, and there shouldn't be a path of normal termination that doesn't call `stop`. 

Also, they weren't daemon threads before (but of course that doesn't mean we shouldn't apply the change if it was necessary).

**Comment by kkonstantine** [comment]
The build was successful. It failed after the tests, being unable to generate the JUnit test result report.

**Comment by ewencp** [comment]
LGTM. Thanks!

**Pull request closed without merging**